File: Leetcode/cpp/sol344.cpp
// https://leetcode.com/problems/reverse-string/description/

class Solution {
public:
    // void swap(vector<char>& s, int i, int j){
    //     s[i]^=s[j];
    //     s[j]^=s[i];
    //     s[i]^=s[j];
    // }

    void reverseString(vector<char>& s) {
        int length = s.size();
        for(int i = 0; i<length/2;i++){
            swap(s[i], s[length-1-i]);
        }
    }
};


class Solution {
public:
    void swap(vector<char>& s, int i, int j){
        s[i]=s[i]+s[j];
        s[j]=s[i]-s[j];
        s[i]=s[i]-s[j];
    }

    void reverseString(vector<char>& s) {
        int length = s.size();
        for(int i = 0; i<length/2;i++){
            swap(s, i, length-1-i);
        }
    }
};

File: Leetcode/python/sol344.py

# https://leetcode.com/problems/reverse-string/description/

class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        def helper(s, left=0, right=None):
            if right is None:
                right = len(s)-1
            if left >= right:
                return
            s[left], s[right] = s[right], s[left]
            helper(s, left+1, right-1)

        helper(s)
